### PR TITLE
Tyr: delete instance API

### PR DESCRIPTION
**Pull request by francois-vincent**
migration tested

### CODE REVIEW

File: source/tyr/migrations/versions/d1d12707b76_add_discarded_instance.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+# revision identifiers, used by Alembic.
+revision = 'd1d12707b76'
+down_revision = '4fa7f3f529e'
+
+from alembic import op
+import sqlalchemy as sa
+
+
+def upgrade():
+    ### commands auto generated by Alembic - please adjust! ###
```
Comment (antoine-de) [inline source/tyr/migrations/versions/d1d12707b76_add_discarded_instance.py]: you should remove those generated comments

File: source/tyr/tests/integration/instance_test.py [hunk 2]
```diff
@@ -42,6 +43,18 @@ def test_get_instances(create_instance):
...
+
+def test_get_instance(create_instance):
+    resp = api_get('/v0/instances/fr')
+    assert len(resp) == 1
+    assert resp[0]['name'] == 'fr'
+    assert resp[0]['id'] == create_instance
+    resp = api_get('/v0/instances/{}'.format(create_instance))
+    assert len(resp) == 1
+    assert resp[0]['name'] == 'fr'
+    assert resp[0]['id'] == create_instance
```
Comment (antoine-de) [inline source/tyr/tests/integration/instance_test.py:55]: :+1:

**Comment by antoine-de** [comment]
general remark: for Navitia it's important to have a good PR name because it is used in the generated changelog.

You should first put the general component changed (Tyr for your PR), and then a short description.

And why did you open a new PR ? you could have just pushed the new changes in the previously opened ( #1610)

**Comment by francois-vincent** [comment]
@antoine-de  ok for the PR name
I changed to a new PR, because of the accumulation of commits

File: source/navitiacommon/navitiacommon/models.py [hunk 3]
```diff
@@ -329,10 +330,23 @@ def last_datasets(self, nb_dataset=1):
...
         return res
 
+    @classmethod
+    def get_from_id_or_name(cls, id=None, name=None):
+        if id:
+            return cls.query.get_or_404(id)
+        elif name:
+            return cls.query_existing().filter_by(name=name).first_or_404()
+        else:
+            return {'error': 'instance is required'}, 400
```
Comment (antoine-de) [inline source/navitiacommon/navitiacommon/models.py]: is this done in other API ? I'm not a fan of the fact that you need to explicitly get returned tuple from the calling function and then treat it as an error.

wouldn't it be better to throw an exception (an exception that will end with a 404) ?

@kinnou02 what do you think ?

**Comment by antoine-de** [comment]
it's not important, but for the accumulation of commit, you can just squash them and push force on your branch

**Comment by antoine-de** [comment]
Jenkins, test this please

File: source/tyr/migrations/versions/d1d12707b76_add_discarded_instance.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    to manage instance deletion.
+
+Revision ID: d1d12707b76
+Revises: 15249a2aa765
+Create Date: 2016-05-31 12:11:00.704902
+
+"""
+
+revision = 'd1d12707b76'
+down_revision = '4fa7f3f529e'
```
Comment (pbougue) [inline source/tyr/migrations/versions/d1d12707b76_add_discarded_instance.py]: you have to update this

File: source/tyr/tyr/resources.py [hunk 5]
```diff
@@ -289,18 +290,27 @@ def get(self, id=None, name=None):
...
-    def put(self, id=None, name=None):
-        if id:
-            instance = models.Instance.query.get_or_404(id)
-        elif name:
-            instance = models.Instance.query.filter_by(name=name).first_or_404()
-        else:
-            return ({'error': 'instance is required'}, 400)
+    def delete(self, id=None, name=None):
+        instance = models.Instance.get_from_id_or_name(id, name)
+        if isinstance(instance, tuple):
```
Comment (xlqian) [inline source/tyr/tyr/resources.py]: in fact, you decide if `instance` is good by testing the type of the returned variable.... @antoine-de  is right, please  manage error cases  of `get_from_id_or_name` by exception...

**Pull request merged**